vendor/k8s.io/kubernetes/pkg/controller/daemon/patch_nodeselector.go

```diff
@@ -0,0 +1,73 @@
+package daemon
+
+import (
+	"k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	appsinformers "k8s.io/client-go/informers/apps/v1beta1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	extensionsinformers "k8s.io/client-go/informers/extensions/v1beta1"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+func NewNodeSelectorAwareDaemonSetsController(defaultNodeSelector string, namepaceInformer coreinformers.NamespaceInformer, daemonSetInformer extensionsinformers.DaemonSetInformer, historyInformer appsinformers.ControllerRevisionInformer, podInformer coreinformers.PodInformer, nodeInformer coreinformers.NodeInformer, kubeClient clientset.Interface) (*DaemonSetsController, error) {
+	controller, err := NewDaemonSetsController(daemonSetInformer, historyInformer, podInformer, nodeInformer, kubeClient)
+	if err != nil {
+		return controller, err
+	}
+	controller.namespaceLister = namepaceInformer.Lister()
+	controller.namespaceStoreSynced = namepaceInformer.Informer().HasSynced
+	controller.defaultNodeSelectorString = defaultNodeSelector
+	if len(controller.defaultNodeSelectorString) > 0 {
+		controller.defaultNodeSelector, err = labels.Parse(controller.defaultNodeSelectorString)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return controller, nil
+}
+
+func (dsc *DaemonSetsController) namespaceNodeSelectorMatches(node *v1.Node, ds *extensions.DaemonSet) (bool, error) {
+	if dsc.namespaceLister == nil {
+		return true, nil
+	}
+
+	// this is racy (different listers) and we get to choose which way to fail.  This should requeue.
+	ns, err := dsc.namespaceLister.Get(ds.Namespace)
+	if apierrors.IsNotFound(err) {
+		return false, err
+	}
+
+	return dsc.nodeSelectorMatches(node, ns), nil
+}
+
+func (dsc *DaemonSetsController) nodeSelectorMatches(node *v1.Node, ns *v1.Namespace) bool {
+	projectNodeSelector, ok := ns.Annotations["openshift.io/node-selector"]
+	switch {
+	case ok:
+		selector, err := labels.Parse(projectNodeSelector)
+		if err == nil {
+			if !selector.Matches(labels.Set(node.Labels)) {
+				return false
+			}
+		}
+	case !ok && len(dsc.defaultNodeSelectorString) > 0:
```

Comment:
@aveshagarwal I think the case is handled right here.

Reply:
@deads so it seems that it might work with openshift's project node selector which is `nodeenv` but it would not work with upstream podnodeselector which specifies cluster level default selector in its config file, right?

Reply:
> @deads so it seems that it might work with openshift's project node selector which is nodeenv but it would not work with upstream podnodeselector which specifies cluster level default selector in its config file, right?

Give me a link and I can plumb that through.  In concept this is still ok.

Reply:
https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/plugin/pkg/admission/podnodeselector/admission.go#L257

Reply:
> ttps://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/plugin/pkg/admission/podnodeselector/admission.go#L257

So it's a semantically different key in a single shared map separated only by namespace validation that prevents capital letters in namespace names?  That's really weird.  Plumb-able, but really weird.

Reply:
Sorry for being dense, but I am not sure I understand the weirdness, so If you could explain it more, I would be happy to fix it :-).

Reply:
> So it's a semantically different key in a single shared map separated only by namespace validation that prevents capital letters in namespace names? That's really weird.

Using one map with two semantically different sets of key/value-pairs is weird.  It would be better to have a separate field on the struct that indicates that something is a default instead of an override.